diff -urN ocsigenserver-2.1-orig/Makefile ocsigenserver-2.1/Makefile
--- ocsigenserver-2.1-orig/Makefile	2012-06-13 20:29:18.442715238 -0400
+++ ocsigenserver-2.1/Makefile	2012-06-13 20:29:39.957714602 -0400
@@ -138,7 +138,7 @@
 	     | sed s%USER%$(OCSIGENUSER)%g \
 	     | sed s%GROUP%$(OCSIGENGROUP)%g \
 	     | sed s%_COMMANDPIPE_%$(COMMANDPIPE)%g \
-	     > $(TEMPROOT)/etc/logrotate.d/$(PROJECTNAME) }
+	     > $(TEMPROOT)/etc/logrotate.d/$(PROJECTNAME)
 
 ###
 
@@ -152,4 +152,4 @@
 
 .PHONY: depend
 depend:
-	${MAKE} -C src depend
\ No newline at end of file
+	${MAKE} -C src depend
