diff -urN ocsigenserver-2.0.3-orig/Makefile ocsigenserver-2.0.3/Makefile
--- ocsigenserver-2.0.3-orig/Makefile	2012-03-01 22:57:29.944048606 -0500
+++ ocsigenserver-2.0.3/Makefile	2012-03-01 23:16:41.229163168 -0500
@@ -93,7 +93,8 @@
 	$(INSTALL) -d -m 755 -o ${OCSIGENUSER} -g ${OCSIGENGROUP} \
 	  $(TEMPROOT)$(STATICPAGESDIR)/ocsigenstuff
 	$(INSTALL) -m 644 -o ${OCSIGENUSER} -g ${OCSIGENGROUP} \
-	  local/var/www/ocsigenstuff/* $(TEMPROOT)$(STATICPAGESDIR)/ocsigenstuff
+	local/var/www/ocsigenstuff/*.png local/var/www/ocsigenstuff/*.css \
+	$(TEMPROOT)$(STATICPAGESDIR)/ocsigenstuff
 	$(INSTALL) -d -m 755 $(TEMPROOT)$(MANDIR)
 	$(INSTALL) -m 644 src/files/${PROJECTNAME}.1 $(TEMPROOT)$(MANDIR)
 
@@ -137,7 +138,7 @@
 	     | sed s%USER%$(OCSIGENUSER)%g \
 	     | sed s%GROUP%$(OCSIGENGROUP)%g \
 	     | sed s%_COMMANDPIPE_%$(COMMANDPIPE)%g \
-	     > $(TEMPROOT)/etc/logrotate.d/$(PROJECTNAME); }
+	     > $(TEMPROOT)/etc/logrotate.d/$(PROJECTNAME)
 
 ###
 
