diff -urN mediastreamer-2.11.1-orig/tools/Makefile.am mediastreamer-2.11.1/tools/Makefile.am
--- mediastreamer-2.11.1-orig/tools/Makefile.am	2015-04-09 22:55:41.329960906 -0400
+++ mediastreamer-2.11.1/tools/Makefile.am	2015-04-09 22:55:21.853958968 -0400
@@ -42,7 +42,8 @@
 	$(VP8_LIBS) \
 	$(VIDEO_LIBS) \
 	$(PCAP_LIBS) \
-	$(POLARSSL_LIBS)
+	$(POLARSSL_LIBS) \
+	-lXext
 
 if ORTP_ENABLED
 if MS2_FILTERS
