diff -urN mediastreamer-2.10.0-orig/src/Makefile.am mediastreamer-2.10.0/src/Makefile.am
--- mediastreamer-2.10.0-orig/src/Makefile.am	2014-09-28 17:47:27.454705851 -0400
+++ mediastreamer-2.10.0/src/Makefile.am	2014-09-28 17:48:04.610791912 -0400
@@ -425,8 +425,8 @@
 
 yuv2rgb.fs.h: yuv2rgb.fs
 	cd $(abs_srcdir) && \
-	xxd -i yuv2rgb.fs | sed s/}\;/,0x00}\;/ > $(abs_builddir)/yuv2rgb.fs.h
+	xxdi.pl yuv2rgb.fs | sed s/}\;/0x00}\;/ > $(abs_builddir)/yuv2rgb.fs.h
 
 yuv2rgb.vs.h: yuv2rgb.vs
 	cd $(abs_srcdir) && \
-	xxd -i yuv2rgb.vs | sed s/}\;/,0x00}\;/ > $(abs_builddir)/yuv2rgb.vs.h
+	xxdi.pl yuv2rgb.vs | sed s/}\;/0x00}\;/ > $(abs_builddir)/yuv2rgb.vs.h
