diff -urN julia-9999-orig/Make.inc julia-9999/Make.inc
--- julia-9999-orig/Make.inc	2012-11-23 12:44:57.277643574 -0500
+++ julia-9999/Make.inc	2012-11-23 12:45:47.984684448 -0500
@@ -17,13 +17,6 @@
 JL_PRIVATE_LIBDIR = lib/julia
 JL_LIBDIR = lib
 
-# If we're on debian, default to arch-dependent library dirs
-ifeq ($(USE_DEBIAN), 1)
-MULTIARCH = $(shell gcc -print-multiarch)
-JL_PRIVATE_LIBDIR = lib/$(MULTIARCH)/julia
-JL_LIBDIR = lib/$(MULTIARCH)/
-endif
-
 OS = $(shell uname)
 ARCH = $(shell uname -m)
 
