diff -urN julia-9999-orig/Makefile julia-9999/Makefile
--- julia-9999-orig/Makefile	2012-11-23 12:43:06.502553282 -0500
+++ julia-9999/Makefile	2012-11-23 12:43:45.162586240 -0500
@@ -33,7 +33,7 @@
 # use sys.ji if it exists, otherwise run two stages
 $(BUILD)/$(JL_PRIVATE_LIBDIR)/sys.ji: VERSION base/*.jl $(BUILD)/share/julia/helpdb.jl
 	$(QUIET_JULIA) cd base && \
-	(test -f $(BUILD)/$(JL_PRIVATE_LIBDIR)/sys.ji || $(JULIA_EXECUTABLE) -bf sysimg.jl) && $(JULIA_EXECUTABLE) -f sysimg.jl || echo "Note: this error is usually fixed by running 'make cleanall'."
+	(test -f $(BUILD)/$(JL_PRIVATE_LIBDIR)/sys.ji || LD_LIBRARY_PATH=../usr/lib/julia:../usr/lib $(JULIA_EXECUTABLE) -bf sysimg.jl) && LD_LIBRARY_PATH=../usr/lib/julia:../usr/lib $(JULIA_EXECUTABLE) -f sysimg.jl || echo "Note: this error is usually fixed by running 'make cleanall'."
 
 ifeq ($(OS), WINNT)
 OPENBLASNAME=openblas-r0.1.1
