diff -urN gmpy2-2.0.0b1-orig/setup.py gmpy2-2.0.0b1/setup.py
--- gmpy2-2.0.0b1-orig/setup.py	2012-05-10 10:59:24.715024051 -0400
+++ gmpy2-2.0.0b1/setup.py	2012-05-10 11:01:06.149021082 -0400
@@ -58,14 +58,14 @@
             break
     for adir in dirord:
         lookin = '%s/lib' % adir
-        if os.path.isfile(lookin + '/lib' + mplib + '.a'):
+        if os.path.isfile(lookin + '/lib' + mplib + '.so'):
             libdirs = [lookin]
             if local_dir and lookin.startswith(local_dir):
                 rundirs = [lookin]
             # Verify that MPFR and MPC exist in the same directory
-            if not os.path.isfile(lookin + '/libmpfr.a'):
+            if not os.path.isfile(lookin + '/libmpfr.so'):
                 use_mpfr = False
-            if not os.path.isfile(lookin + '/libmpc.a'):
+            if not os.path.isfile(lookin + '/libmpc.so'):
                 use_mpc = False
             break
 
