diff -urN scilab-6.0.0-alpha-2-orig/modules/ast/Makefile.am scilab-6.0.0-alpha-2/modules/ast/Makefile.am
--- scilab-6.0.0-alpha-2-orig/modules/ast/Makefile.am	2015-11-17 22:07:17.103414427 -0500
+++ scilab-6.0.0-alpha-2/modules/ast/Makefile.am	2015-11-17 22:08:03.868803361 -0500
@@ -246,11 +246,8 @@
 src/cpp/analysis/check_sum.cpp \
 src/cpp/analysis/check_____ldivide____.cpp \
 src/cpp/analysis/check_acos.cpp \
-src/cpp/analysis/check_____dottimes____.cpp
-
-if ENABLE_DEBUG
-libsciast_la_SOURCES += src/cpp/types/inspector.cpp
-endif
+src/cpp/analysis/check_____dottimes____.cpp \
+src/cpp/types/inspector.cpp
 
 libsciast_la_CPPFLAGS = \
 	-I$(srcdir)/includes/ast \
