diff -urN scilab-5.4.0-alpha-1-orig/m4/java.m4 scilab-5.4.0-alpha-1/m4/java.m4
--- scilab-5.4.0-alpha-1-orig/m4/java.m4	2012-04-12 16:30:27.528288410 -0400
+++ scilab-5.4.0-alpha-1/m4/java.m4	2012-04-12 16:31:46.167286119 -0400
@@ -823,7 +823,7 @@
     PACKAGE_JAR_FILE=
     found_jar=no
     saved_ac_java_classpath=$ac_java_classpath
-    DEFAULT_JAR_DIR="/usr/share/java /usr/lib/java /usr/share/java /usr/share/java/jar /opt/java/lib /usr/local/java /usr/local/java/jar /usr/local/share/java /usr/local/share/java/jar /usr/local/lib/java $(ls -d /usr/share/java/* 2>/dev/null) $(ls -d /usr/lib64/* 2>/dev/null) $(ls -d /usr/lib/* 2>/dev/null)  $(ls -d /usr/share/*/lib 2>/dev/null)"
+    DEFAULT_JAR_DIR="$(ls -d /usr/share/*/lib 2>/dev/null)"
     for jardir in "`pwd`/thirdparty" "`pwd`/jar" $DEFAULT_JAR_DIR "$_user_libdir"; do
       for jar in "$jardir/$1.jar" "$jardir/lib$1.jar" "$jardir/lib$1-java.jar" "$jardir/$1*.jar"; do
 #    jar=`echo $jar|sed -e 's/ /\\ /'`
