--- m4/compiler-orig.m4	2013-09-19 23:54:02.462761181 -0400
+++ m4/compiler.m4	2013-09-19 23:55:54.927772372 -0400
@@ -30,36 +30,41 @@
 AC_DEFUN([AC_CHECK_STDCPP_STATIC],[
 
 AC_ARG_ENABLE(static-system-lib,
-    AC_HELP_STRING([--disable-static-system-lib],[Disable the usage of the static stdc++ and libgcc libs]))
-
-# Check if -static-libstdc++ and -static-libgcc are supported.
-saved_LDFLAGS="$LDFLAGS"
-
-compiler_manage_static_libs=no
-USE_STATIC_SYSTEM_LIB=yes
-
-LDFLAGS="$LDFLAGS -static-libstdc++ -static-libgcc"
-AC_MSG_CHECKING([whether g++ accepts -static-libstdc++ -static-libgcc])
-AC_LANG_PUSH(C++)
-AC_LINK_IFELSE([
-#if (__GNUC__ < 4) || (__GNUC__ == 4 && __GNUC_MINOR__ < 5)
-#error -static-libstdc++ not implemented
-#endif
-int main() {}],
-  [AC_MSG_RESULT([yes]); compiler_manage_static_libs=yes],
-  [AC_MSG_RESULT([no])])
-AC_LANG_POP(C++)
-
-LDFLAGS="$saved_LDFLAGS"
-if test "x$compiler_manage_static_libs" == "xno" -a "x$enable_static_system_lib" != "xno"; then
-   AC_MSG_ERROR([The $CC compiler does not support the options -static-libstdc++/-static-libgcc. Update your compiler or use the option --disable-static-system-lib])
+    AC_HELP_STRING([--disable-static-system-lib],[Disable the usage of the static stdc++ and libgcc libs]),
+    [enable_static_system_lib=$enableval],
+    [enable_static_system_lib='yes']
+    )
+
+if test "x$enable_static_system_lib" != "xno"; then
+
+	# Check if -static-libstdc++ and -static-libgcc are supported.
+	saved_LDFLAGS="$LDFLAGS"
+
+	compiler_manage_static_libs=no
+	USE_STATIC_SYSTEM_LIB=yes
+	
+	LDFLAGS="$LDFLAGS -static-libstdc++ -static-libgcc"
+	AC_MSG_CHECKING([whether g++ accepts -static-libstdc++ -static-libgcc])
+	AC_LANG_PUSH(C++)
+	AC_LINK_IFELSE([
+	#if (__GNUC__ < 4) || (__GNUC__ == 4 && __GNUC_MINOR__ < 5)
+	#error -static-libstdc++ not implemented
+	#endif
+	int main() {}],
+	  [AC_MSG_RESULT([yes]); compiler_manage_static_libs=yes],
+	  [AC_MSG_RESULT([no])])
+	AC_LANG_POP(C++)
+	
+	LDFLAGS="$saved_LDFLAGS"
+	if test "x$compiler_manage_static_libs" == "xno"; then
+	   AC_MSG_ERROR([The $CC compiler does not support the options -static-libstdc++/-static-libgcc. Update your compiler or use the option --disable-static-system-lib])
+	fi
+	
+	if test "x$compiler_manage_static_libs" == "xno"; then
+	   echo "-static-libstdc++ -static-libgcc unsupported by the compiler. Skipped"
+	   USE_STATIC_SYSTEM_LIB=no
+	fi
 fi
-
-if test "x$compiler_manage_static_libs" == "xno"; then
-   echo "-static-libstdc++ -static-libgcc unsupported by the compiler. Skipped"
-   USE_STATIC_SYSTEM_LIB=no
-fi
-
 AM_CONDITIONAL(USE_STATIC_SYSTEM_LIB, test "$USE_STATIC_SYSTEM_LIB" == yes)
 
 ])
