diff -urN flexdock-orig/src/java/drag/org/flexdock/docking/drag/outline/xlib/XlibRubberBand.java flexdock/src/java/drag/org/flexdock/docking/drag/outline/xlib/XlibRubberBand.java
--- flexdock-orig/src/java/drag/org/flexdock/docking/drag/outline/xlib/XlibRubberBand.java	2011-11-24 16:17:08.000000000 -0500
+++ flexdock/src/java/drag/org/flexdock/docking/drag/outline/xlib/XlibRubberBand.java	2011-12-20 18:33:25.794076694 -0500
@@ -17,7 +17,7 @@
  *
  */
 public class XlibRubberBand extends RubberBand {
-    private static final String NATIVE_RESOURCE_PATH = "org/flexdock/docking/drag/outline/xlib/";
+    private static final String NATIVE_RESOURCE_PATH = "/usr/lib/flexdock";
     private static final String NATIVE_RESOURCE = "libRubberBand.so";
     private static final String NATIVE_RESOURCE_START = "libRubberBand";
     private static final String NATIVE_RESOURCE_END = ".so";
@@ -47,6 +47,7 @@
 
             try {
                 ResourceManager.loadLibrary(lib, resource);
+                System.load(resource);
                 // if the library was successfully loaded, then we don't
                 // need to do anything else.
                 return;
