diff -urN flexdock-orig//src/java/drag/org/flexdock/docking/drag/outline/xlib/XlibRubberBand.java flexdock/src/java/drag/org/flexdock/docking/drag/outline/xlib/XlibRubberBand.java
--- flexdock-orig//src/java/drag/org/flexdock/docking/drag/outline/xlib/XlibRubberBand.java	2005-08-27 04:51:10.000000000 -0400
+++ flexdock/src/java/drag/org/flexdock/docking/drag/outline/xlib/XlibRubberBand.java	2010-09-14 22:54:12.143000075 -0400
@@ -17,7 +17,7 @@
  *
  */
 public class XlibRubberBand extends RubberBand {
-    private static final String NATIVE_RESOURCE_PATH = "org/flexdock/docking/drag/outline/xlib/";
+    private static final String NATIVE_RESOURCE_PATH = "/usr/lib/flexdock";
     private static final String NATIVE_RESOURCE = "libRubberBand.so";
     private static final String NATIVE_RESOURCE_START = "libRubberBand";
     private static final String NATIVE_RESOURCE_END = ".so";
@@ -47,6 +47,7 @@
 	        
 	        try {
 	            ResourceManager.loadLibrary(lib, resource);
+		    System.load(resource);
 	            // if the library was successfully loaded, then we don't 
 	            // need to do anything else.
 	            return;
