--- chromaprint-1.0.orig/src/ext/ffmpeg_decoder.h	2013-02-16 17:49:06.799592088 +0100
+++ chromaprint-1.0/src/ext/ffmpeg_decoder.h	2013-02-16 17:49:54.037590369 +0100
@@ -28,6 +28,10 @@
 }
 #include "audio_consumer.h"
 
+#ifndef AVCODEC_MAX_AUDIO_FRAME_SIZE
+#  define AVCODEC_MAX_AUDIO_FRAME_SIZE 192000
+#endif
+
 class Decoder
 {
 public:
@@ -124,7 +128,7 @@
 		return false;
 	}
 
-	if (avcodec_open(m_codec_ctx, codec) < 0) {
+	if (avcodec_open2(m_codec_ctx, codec, NULL) < 0) {
 		m_error = "Couldn't open the codec.";
 		return false;
 	}
