diff -urN zarith-1.2-orig/project.mak zarith-1.2/project.mak
--- zarith-1.2-orig/project.mak	2013-06-12 00:10:56.054290489 -0400
+++ zarith-1.2/project.mak	2013-06-12 00:09:36.500292819 -0400
@@ -109,7 +109,7 @@
 
 ifeq ($(INSTMETH),findlib)
 install:
-	$(OCAMLFIND) install -destdir "$(INSTALLDIR)" zarith META $(TOINSTALL) dllzarith.dll
+	$(OCAMLFIND) install -destdir "$(INSTALLDIR)" zarith META $(TOINSTALL) dllzarith.$(DLLSUFFIX)
 
 uninstall:
 	$(OCAMLFIND) remove -destdir "$(INSTALLDIR)" zarith
