diff -urN pjproject-2.1.0-orig/pjmedia/src/pjmedia/ffmpeg_util.c pjproject-2.1.0/pjmedia/src/pjmedia/ffmpeg_util.c
--- pjproject-2.1.0-orig/pjmedia/src/pjmedia/ffmpeg_util.c	2013-03-24 12:11:22.085580704 -0400
+++ pjproject-2.1.0/pjmedia/src/pjmedia/ffmpeg_util.c	2013-03-24 12:13:14.020591842 -0400
@@ -33,6 +33,9 @@
                               	 MAKE_VER(mj,mn,mi))
 
 
+#ifndef PIX_FMT_GBR24P
+#define PIX_FMT_GBR24P PIX_FMT_GBRP
+#endif
 
 /* Conversion table between pjmedia_format_id and PixelFormat */
 static const struct ffmpeg_fmt_table_t
