diff -urN pjproject-2.1.0-orig/build.mak.in pjproject-2.1.0/build.mak.in
--- pjproject-2.1.0-orig/build.mak.in	2013-03-25 21:00:56.668643948 -0400
+++ pjproject-2.1.0/build.mak.in	2013-03-25 21:01:04.962644774 -0400
@@ -26,8 +26,6 @@
 export APP_THIRD_PARTY_LIBS := -lresample $(APP_THIRD_PARTY_LIBS)
 export APP_THIRD_PARTY_LIB_FILES := $(PJ_DIR)/third_party/lib/libresample.$(SHLIB_SUFFIX).$(PJ_VERSION_MAJOR) $(PJ_DIR)/third_party/lib/libresample.$(SHLIB_SUFFIX) $(APP_THIRD_PARTY_LIB_FILES)
 else
-export APP_THIRD_PARTY_LIBS := -lresample-$(TARGET_NAME) $(APP_THIRD_PARTY_LIBS)
-export APP_THIRD_PARTY_LIB_FILES := $(PJ_DIR)/third_party/lib/libresample-$(LIB_SUFFIX) $(APP_THIRD_PARTY_LIB_FILES)
 endif
 
 ifneq (@ac_no_gsm_codec@,1)
diff -urN pjproject-2.1.0-orig/pjmedia/src/pjmedia/resample_libsamplerate.c pjproject-2.1.0/pjmedia/src/pjmedia/resample_libsamplerate.c
--- pjproject-2.1.0-orig/pjmedia/src/pjmedia/resample_libsamplerate.c	2013-03-25 21:00:56.718643953 -0400
+++ pjproject-2.1.0/pjmedia/src/pjmedia/resample_libsamplerate.c	2013-03-25 21:01:29.310647197 -0400
@@ -33,7 +33,7 @@
 
 #if PJMEDIA_RESAMPLE_IMP==PJMEDIA_RESAMPLE_LIBSAMPLERATE
 
-#include "../../third_party/libsamplerate/src/samplerate.h"
+#include "samplerate.h"
 
 #define THIS_FILE   "resample_libsamplerate.c"
 
